enhancements/microshift/workload-partitioning-in-microshift.md

```diff
@@ -0,0 +1,336 @@
+data = """
+kubelet:
+  reservedSystemCPUs: "0,6,7"
+  cpuManagerPolicy: static
+  cpuManagerPolicyOptions:
+    full-pcpus-only: "true"
+  cpuManagerReconcilePeriod: 5s
+"""
+
+[[customizations.files]]
+path = "/etc/crio/crio.conf.d/20-microshift-wp.conf"
+data = """
+[crio.runtime]
+infra_ctr_cpuset = "0,6,7"
+
+[crio.runtime.workloads.management]
+activation_annotation = "target.workload.openshift.io/management"
+annotation_prefix = "resources.workload.openshift.io"
+resources = { "cpushares" = 0, "cpuset" = "0" }
+"""
+
+[[customizations.files]]
+path = "/etc/systemd/system/crio.service.d/microshift-cpuaffinity.conf"
+data = """
+[Service]
+CPUAffinity=0,6,7
+"""
+
+[[customizations.files]]
+path = "/etc/systemd/system/microshift.service.d/microshift-cpuaffinity.conf"
+data = """
+[Service]
+CPUAffinity=0,6,7
+"""
+
+
+[[customizations.files]]
+path = "/etc/kubernetes/openshift-workload-pinning"
+data = """
+{
+  "management": {
+    "cpuset": "0,6,7"
+  }
+}
+"""
+
+```
+
+##### RPM
+1. User creates configuration files with the workload partitioning configuration
+1. reboot the host
+
+### API Extensions
+
+Following API extensions are expected:
+- A passthrough from MicroShift's config to Kubelet config.
+
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+Purely MicroShift enhancement.
+
+### Implementation Details/Notes/Constraints
+
+#### Kubelet configuration 
+- instructs kubelet to modify the node resource with the capacity and allocatable CPUs.
+  
+  add to the file /etc/kubernetes/openshift-workload-pinning
+  ```json
+  {
+    "management": {
+      "cpuset": "0,6,7"
+    }
+  }
+  ```
+- microshift passthrough configuration for kubelet
+
+  add to file /etc/microshift/config.yaml 
+  ```yaml
+  kubelet:
+    reservedSystemCPUs: 0,6,7
+    cpuManagerPolicy: static
+    cpuManagerPolicyOptions:
+      full-pcpus-only: "true"
+    cpuManagerReconcilePeriod: 5s    
+  ```
+
+#### CRI-O configuration
+
+- add this drop-in to CRI-I with the following example:
+  ```ini
+  [crio.runtime]
+  infra_ctr_cpuset = "0,6,7"
+
+  [crio.runtime.workloads.management]
+  activation_annotation = "target.workload.openshift.io/management"
+  annotation_prefix = "resources.workload.openshift.io"
+  resources = { "cpushares" = 0, "cpuset" = "0,6,7" }
+  ```
+
+- add systemd drop-in for crio
+
+  in the file /etc/systemd/system/crio.service.d/microshift-cpuaffinity.conf
+  ```ini
+  [Service]
+  CPUAffinity=0,6,7
+  ```
+
+####  Admission Webhook 
+Admission Webhook to update the Control Plane POD Annotations  for the CRIO
+Introduce cpuset profiles to Microshift configuration
+
+
+#### Microshift Control Plane cpu pinning
+MicroShift runs as a single systemd unit. The main binary embeds as goroutines only those services strictly necessary to bring up a *minimal Kubernetes/OpenShift control and data plane*. 
```

Comment:
Actually there's `microshift-etcd.scope` unit which is created at runtime. Not sure if it will inherit the CPUAffinity of `microshift.service` or not

Reply:
I've tested it , it does inherit it.

Reply:
# taskset -p $(pidof microshift)
pid 71033's current affinity mask: 1

# taskset -p $(pidof microshift-etcd)
pid 71046's current affinity mask: 1

# taskset -p $(pidof crio)
pid 1095's current affinity mask: 1